### PR TITLE
History service

### PR DESCRIPTION
**Pull request by Domszy**
# Description

Allows the user to access past attempts on PeerPrep. It also showcases the number of easy, medium and hard questions the user have done in `MetricComponent` and also a table of the attempts made by the user in `PreviousQuestionDone`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by benedictchuajj** [comment]
Seems like there's this 'bug' where the first user that closes the room (and is returned home automatically via close room) doesn't have their history updated. 

However, doesn't seem like an important issue since it's updated once we refresh the browser.

File: frontend/src/components/MetricComponent.js [hunk 1]
```diff
@@ -2,10 +2,58 @@ import { PieChart } from '@mui/x-charts/PieChart'
...
+
+  // const hasQuestionDone = () => {
+  //   if (previousQuestions === undefined) {
+  //     return false
+  //   } else {
+  //     return true
+  //   }
+  // }
+
+  const getEasyQuestion = () => {
```
Comment (benedictchuajj) [inline frontend/src/components/MetricComponent.js:21]: Logic looks similar to the ones below. Perhaps you can abstract this out.

File: frontend/src/components/PreviousQuestionEditor.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import React from 'react'
+import CodeMirror from '@uiw/react-codemirror'
+import { materialLightInit } from '@uiw/codemirror-theme-material'
+import { python } from '@codemirror/lang-python'
+import { java } from '@codemirror/lang-java'
+import { cpp } from '@codemirror/lang-cpp'
+import { Select, MenuItem, InputLabel, FormControl, Card } from '@mui/material'
+
+const PreviousQuestionEditor = ({ code, programmingLanguage }) => {
+  // const EditorViewCodeMirror = new EditorView({
```
Comment (benedictchuajj) [inline frontend/src/components/PreviousQuestionEditor.js:10]: Do you need this?

File: frontend/src/components/MetricComponent.js [hunk 3]
```diff
@@ -2,10 +2,58 @@ import { PieChart } from '@mui/x-charts/PieChart'
...
+  selectPreviousQuestions
+} from '../redux/UserSlice'
 
-const MetricComponent = ({ easyCount, normalCount, hardCount }) => {
-  const totalQuestions = easyCount + normalCount + hardCount
-  const hasQuestionDone = totalQuestions > 0
+const MetricComponent = () => {
+  const previousQuestions = useSelector(selectPreviousQuestions)
+
+  // const hasQuestionDone = () => {
```
Comment (benedictchuajj) [inline frontend/src/components/MetricComponent.js:13]: Do you need this?

**Review by benedictchuajj — APPROVED** [review]
Have small nits but LGTM

**Review by KT27Learn — APPROVED** [review]
LGTM

**Comment by benedictchuajj** [comment]
Merged master and resolved conflict. 

Will merge in this PR to master first

**Pull request merged**